pkg/binding/controller.go

```diff
@@ -99,17 +99,27 @@ type Controller struct {
 func NewController(parentLogger logr.Logger, wdsRestConfig *rest.Config, imbsRestConfig *rest.Config,
 	wdsName string, allowedGroupsSet sets.Set[string]) (*Controller, error) {
 
-	dynamicClient, err := dynamic.NewForConfig(wdsRestConfig)
+	kubernetesClient, err := kubernetes.NewForConfig(wdsRestConfig)
 	if err != nil {
 		return nil, err
 	}
 
-	kubernetesClient, err := kubernetes.NewForConfig(wdsRestConfig)
+	extClient, err := apiextensionsclientset.NewForConfig(wdsRestConfig)
 	if err != nil {
 		return nil, err
 	}
 
-	extClient, err := apiextensionsclientset.NewForConfig(wdsRestConfig)
+	burst, err := countGVRsViaAggregatedDiscovery(rest.CopyConfig(wdsRestConfig)) // use # of GVRs as burst is tested to be working well
+	if err != nil {
+		parentLogger.Error(err, "Not able to count GVRs via aggregated discovery") // but we can still continue
+	}
+	wdsRestConfig.Burst = adjustBurstIfNecessary(burst)
+	wdsRestConfig.QPS = adjustQPSIfNecessary(float32(burst) / 4)
```

Comment:
I rewrote this section of code, and renamed the functions. Does it look better now? BTW what is the exact 'side-effect' you mentioned above?

Reply:
The side-effects are still there. I am talking about modifying the struct that the pointer parameter `wdsRestConfig` points to.

Reply:
Still not following. Which line modifies the struct `wdsRestConfig`, as a pointer parameter (or in other words as 'an input data structure')? I thought the modifications are explicit and not modifying as any input data structure in current lines 118 and 119.

Reply:
I finally see the point (it would have been helpful in the first comment to say the input data structure **of `NewController`**). But I do have a different view on this issue and would like to discuss as follows.

My view is that if a caller passes a pointer to a callee, the caller should be 'prepared' that the pointed content may be modified. If the caller doesn't want the possible modification, than pass a copy or value instead. In other words, if a pointer is passed, than permission of modification is implicit.

Is there some established convention (pls do let me know if there is)? Or this view can vary from person to person?

Reply:
I am not aware of a universal convention about this for programming in Go. However, we are talking here about a more specific subject matter: setting up controllers in Kubernetes code. The convention in Kubernetes code is that the consumer of a REST config does not side-effect the copy given unless there is something explicit in the func's name or comments about that. The consumer can make a copy for its own purpose, if necessary.